sched/semaphore/sem_recover.c

```diff
@@ -100,7 +101,39 @@ void nxsem_recover(FAR struct tcb_s *tcb)
        * place.
        */
 
-      atomic_fetch_add(NXSEM_COUNT(sem), 1);
+      if (NXSEM_IS_MUTEX(sem))
+        {
+          FAR dq_entry_t *wtcb;
+          uint32_t mholder;
+          uint32_t blocks_mask = ~NXSEM_MBLOCKS_BIT;
+
+          /* The TID of the mutex holder is correct but we need to
+           * update the blocking bit. Count the remaining tcbs in the
+           * blocking list (ignoring the one that is being cancelled).
+           */
+
+          for (wtcb = dq_peek(SEM_WAITLIST(sem)); wtcb; wtcb = dq_next(wtcb))
```

Comment:
In this case tcb is waiting for this mutex. The only thing that needs to be done is to remove it from the SEM_WAITLIST. If the waitlist becomes empty, then clear block mask ?

Reply:
Yes. But the issue here is, that I can't remove it from the SEM_WAITLIST here. It is done afterwards. So the loop above does exactly that, it checks if there will be other tasks left after this upcoming removal from the waitlist.

Reply:
I think you are mixing up holder and waiter lists now. For a mutex only a single entry can EVER be in the holder list, for counting semaphores this is different.

However, for waiter list there can be multiple entries.

What I said above is correct, the tcb must be removed from the semaphores waiter list here. Otherwise it stays there forever.

It also looks like for semaphores there is a bug, the tcb being removed stays in the wait list forever...

Reply:
I can find two places where nodes are removed from SEM_WAITLIST:
https://github.com/apache/nuttx/blob/8832136b69865d54124c9a48d3a87c29ecca152c/sched/semaphore/sem_waitirq.c#L98)

https://github.com/apache/nuttx/blob/8832136b69865d54124c9a48d3a87c29ecca152c/sched/semaphore/sem_post.c#L149

So yes, it does look like we leave the dying tcb in the semaphore's waiter list for eternity.

Reply:
Yes, I was reading your comment with eyes crossed. There indeed seems to be an already existing bug in here. The counting semaphores should have been removed from the waiting list just the same. I was copying the existing functionality and didn't just notice that.

I'll make a separate patch fixing this, and add to this same PR.

Reply:
Check that nxsem_wait_irq is not called by the task exit logic, I forgot to verify that

Reply:
Decided to fix it right away. Tested again MPFS SMP / CONFIG_BUILD_FLAT and rv-virt:nsh, also ostest on both. Should be fine...